### PR TITLE
Implementation for JetHT validation

### PR DESCRIPTION
**Pull request by jusaviin**
This is the first implementation of the JetHT validation to the new allInOneConfig. All the tests I made locally work, so I think that the code is ready to be pushed to the common framework.

### CODE REVIEW

File: Alignment/OfflineValidation/interface/SmartSelectionMonitor.h [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+class SmartSelectionMonitor {
+  
+public:
+
+  SmartSelectionMonitor(){}
+  ~SmartSelectionMonitor() { }
+
+  typedef __gnu_cxx::hash_map<TString, std::map<TString, TH1*>* > Monitor_t;
```
Comment (mmusich) [inline Alignment/OfflineValidation/interface/SmartSelectionMonitor.h]: I think in the meanwhile this type of container is deprecated in the master of cmssw, shall we try with an STL unordered map?

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+  tracksToken_      (consumes<reco::TrackCollection>(tracksTag_)),
+  triggerTag_       (iConfig.getParameter<edm::InputTag>("triggerResults")), 
+  triggerToken_     (consumes<edm::TriggerResults>(triggerTag_)),
+  printTriggerTable_(iConfig.getUntrackedParameter<int>("printTriggerTable")), 
+  minVtxNdf_        (iConfig.getUntrackedParameter<double>("minVertexNdf")), 
+  minVtxWgt_ (iConfig.getUntrackedParameter<double>("minVertexMeanWeight")),
+  profilePtBorders_ (iConfig.getUntrackedParameter<std::vector<double>>("profilePtBorders")),
+  iovList_ (iConfig.getUntrackedParameter<std::vector<int>>("iovList"))
+{
+   //now do what ever initialization is needed
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]: `usesResource(TFileService::kSharedResource);`
 is missing here as we are using `TFileService`.

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc [hunk 3]
```diff
@@ -0,0 +1,377 @@
...
+}
+
+
+JetHTAnalyzer::~JetHTAnalyzer()
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]: if this is not used, make it `default`.

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc [hunk 4]
```diff
@@ -0,0 +1,377 @@
...
+JetHTAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
+{
+  using namespace edm;
+
+  const double cmToum = 10000;
+
+  edm::ESHandle<TransientTrackBuilder>            theB                ;
+  edm::ESHandle<GlobalTrackingGeometry>           theTrackingGeometry ;
+  iSetup.get<GlobalTrackingGeometryRecord>().get(theTrackingGeometry) ;
+  iSetup.get<TransientTrackRecord>().get("TransientTrackBuilder",theB);
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]: Please use esConsumes here.
For details see:  https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES#Getting_data_from_EventSetup_wit

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+
+  edm::Handle<reco::VertexCollection> vertices; 
+  iEvent.getByToken(pvsToken_, vertices);
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]:
```suggestion
  const auto& vertices = iEvent.get(pvsToken);
```
as per suggestion at https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideEDMGetDataFromEvent#get

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc [hunk 6]
```diff
@@ -0,0 +1,377 @@
...
+  const reco::VertexCollection pvtx  = *(vertices.product())  ;    
+
+  edm::Handle<reco::TrackCollection> tracks; 
+  iEvent.getByToken(tracksToken_, tracks);
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]:
```suggestion
  const auto& tracks = iEvent.get(tracksToken_);
```

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc [hunk 7]
```diff
@@ -0,0 +1,377 @@
...
+        iovString = Form("iov%d-%d",iovList_.at(i-1),iovList_.at(i)-1);
+        break;
+      }
+    }
+  }
+
+  // Print the triggers to console
+  if(printTriggerTable_){
+    edm::Handle<edm::TriggerResults> triggerResults;
+    iEvent.getByToken(triggerToken_, triggerResults);
```
Comment (mmusich) [inline Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]:
```suggestion
    const auto& triggerResults = iEvent.get(triggerToken_);
```

**Review by mmusich — CHANGES_REQUESTED** [review]
Hi @jusaviin,
Thanks for this. I left few comments - only on the cmssw part.

File: Alignment/OfflineValidation/interface/SmartSelectionMonitor.h (hunk 1 above)
Comment (jusaviin) [reply Alignment/OfflineValidation/interface/SmartSelectionMonitor.h]: Yeah, while hash_map works, it gives a deprecation warning during compilation. However, if I just change hash_map to unordered_map, it compiles ok, but none of the histograms get filled. I am not sure what else is needed to make the code work using unordered_map.

File: Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc (hunk 6 above)
Comment (jusaviin) [reply Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]: Later in the code there is a loop which produces tracks with a type reco::TrackRef. There are several constructors for this, but I did not manage to make it work with const auto& tracks. The current code uses the constructor that requires an edm::Handle.

Comment (mmusich) [reply Alignment/OfflineValidation/plugins/JetHTAnalyzer.cc]: You can still use:
```
edm::Handle<reco::TrackCollection> tracks = iEvent.getHandle(tracksToken_);
```

File: Alignment/OfflineValidation/interface/SmartSelectionMonitor.h (hunk 1 above)
Comment (mmusich) [reply Alignment/OfflineValidation/interface/SmartSelectionMonitor.h]: I'll try to see if there's some alternative way.

Comment (jusaviin) [reply Alignment/OfflineValidation/interface/SmartSelectionMonitor.h]: I found the reason why unordered_map did not work. The code was using TString as a key for the unordered_map and the equality operator for TString is not working as one would expect when comparing strings. Namely, if TString one = "myString" and TString two = "myString", (one == two) returns false. This breaks the find-function of the unordered_map and thus the code thought none of the keys it tried to seek existed in the map. The fix was to replace TString with std::string as a key value. After this and some necessary adjustments elsewhere in the code, it now compiles and works without giving the deprecation warning.

Comment (mmusich) [reply Alignment/OfflineValidation/interface/SmartSelectionMonitor.h]: Excellent, thank you

**Review by mmusich — APPROVED** [review]

**Comment by jusaviin** [comment]
The three commits after Marco's approval contain the documentation in a README.md file and a couple of quality of life improvements. I will add the CRAB running option after the main CRAB submission code is merged to the newAllInOneTool.

**Comment by mmusich** [comment]
Is this ready to merge? @connorpa please review

**Comment by connorpa** [comment]
Fine with me

**Pull request merged**